Introduce secure repos

edsko opened the pull request:
So this is the final PR that introduces support for secure repositories.

**NOTE**: Do not merge this, this currently relies on as-yet-unreleased versions of both the `hackage-security` library (it needs the latest `master` from https://github.com/well-typed/hackage-security) as well as the `tar` package (it currently needs the `pr/TarIndex-toList` branch of https://github.com/edsko/tar/). 

Submitting this as a PR right now so that it can be reviewed now and then merged quickly once the other libraries have been released. (Note that we will also want to have come to a decision re dependencies.)

edsko commented:
Note that this currently still hardcodes the keys from the alpha release. That is another thing that will need fixing (each by not having any keys at all or by having the proper root keys if we can get that sorted out before releasing). This PR also means that the HTTPS/curl transport currently available in `cabal-install` cannot be used when using secure repos; that is an open issue (https://github.com/well-typed/hackage-security/issues/42) but not one that necessarily needs to be done before releasing; however, if I get a chance to do it soon then I will.

edsko commented:
(Note also that I have dropped support, including Travis checks, for 7.2.2 from the hackage-security library as we discussed; it's currently failing on 7.4 but this is due to a problem with `http-client`, for which I have submitted a PR https://github.com/snoyberg/http-client/pull/174).

23Skidoo commented on cabal-install/Distribution/Client/Security/HTTP.hs:
```diff
@@ -0,0 +1,307 @@
+--
+-- TODO: This currently uses the lazy bytestring API offered by the HTTP
+-- library. Unfortunately this provides no way of closing the connection when
+-- the callback decides it doens't require any further input. It seems
```
Typo: doens't.

23Skidoo commented on cabal-install/Distribution/Client/Security/HTTP.hs:
```diff
@@ -0,0 +1,307 @@
+  , setOutHandler
+  , setErrHandler
+  , setProxy
+  , request
```
Only `withClient`, `setOutHandler`, `setErrHandler` and `setProxy` seem to be used outside of this module.

edsko replied:
This is a direct copy of the `hackage-security-HTTP` library, that's why. I'll strip it down (though if we merge it with the transport concept already existent in `cabal-install`, this will change anyway).

23Skidoo commented on cabal-install/Distribution/Client/IndexUtils.hs:
```diff
@@ -179,6 +180,9 @@ readRepoIndex verbosity repo mode =
           RepoRemote{..} -> warn verbosity $
                "The package list for '" ++ remoteRepoName repoRemote
             ++ "' does not exist. Run 'cabal update' to download it."
+          RepoSecure{..} -> warn verbosity $
```
May be worth refactoring the error message into a `errorPackageListDoesNotExist` function.

edsko replied:
I had in fact originally done that but when I tried hard to show that the new secure stuff didn't affect any existing code paths I changed it back to this :)

23Skidoo commented on cabal-install/Distribution/Client/IndexUtils.hs:
```diff
@@ -192,6 +196,10 @@ readRepoIndex verbosity repo mode =
              "The package list for '" ++ remoteRepoName repoRemote
           ++ "' is " ++ shows (floor dt :: Int) " days old.\nRun "
           ++ "'cabal update' to get the latest list of available packages."
+        RepoSecure{..} -> warn verbosity $
```
Same here.

23Skidoo commented on cabal-install/Distribution/Client/Security/HTTP.hs:
```diff
@@ -0,0 +1,307 @@
+{-# LANGUAGE CPP #-}
```
This module provides functionality similar to the existing `HttpUtils` module. How difficult would be to merge the two? We can of course do this at a later stage, this is not a requirement for merging.

dcoutts replied:
Yeah I see no problem with merging them into one module.

23Skidoo commented on cabal-install/Distribution/Client/Update.hs:
```diff
@@ -66,3 +69,13 @@ updateRepo transport verbosity _ignoreExpiry repo = case repo of
+            else return Nothing
+    updated <- Sec.uncheckClientErrors $ Sec.checkForUpdates repoSecure ce
+    -- Update cabal's internal index as well so that it's not out of sync
+    -- (If all access to the cache goes through hackage-secury this can go)
```
Typo: hackage-secury

23Skidoo replied:
And yes, would be nice to get rid of the old cache at some point.

23Skidoo commented:
> This PR also means that the HTTPS/curl transport currently available in cabal-install cannot be used when using secure repos

This only concerns fetching the index tarball, not package downloading/uploading? Yes, would be nice to fix, and we should also mention this in the release notes.

23Skidoo commented on cabal-install/Distribution/Client/Setup.hs:
```diff
@@ -387,18 +399,96 @@ instance Monoid GlobalFlags where
+
+-- | Initialize the given secure repo
+--
+-- The security library has its own concept of a "local" repository, distinct
```
This feels a bit confusing. Maybe we can refactor `Repo` to something like `RepoLocal { ... repoSecure :: Maybe SecurityInfo } | RepoRemote { ... repoSecure :: Maybe SecurityInfo}` instead of making `RepoSecure` a top-level constructor? Then instead of `file://...` URLs one would use something like

```
local-repo foo
    path /bar/baz
    secure: True
    [...]
```
in the config file.

edsko replied:
Hmmm, this I am not so keen on. Cabal's concept of a local repository is quite different from any other kind of repository (different format, for one). This would also require a further change to the config file.

23Skidoo replied:
> Cabal's concept of a local repository is quite different from any other kind of repository

Do these differences really affect users that much? I think our users can live with the fact that secure and non-secure local repos have different on-disk format (provided that the error message printed when the `local-repo` is in the wrong format is decent). IMO, `local-repo` is a more logical/discoverable place to specify secure local repos in the config file than `remote-repo` + `file:/`. Also, I don't like the fact that `RemoteRepo` in the code now means "a remote repo, except when it's not" :-).

> This would also require a further change to the config file. 

I think this'd be for the better from the UX point of view, as `local-repo` and `remote-repo` would be nicely symmetric.

23Skidoo replied:
Also pinging @dcoutts, would be interesting to know what he thinks.

edsko replied:
I wonder how many `local-repo`s are out there, if any at all.. Could rename `remote-repo` to simply `repository`.. Anyway, I don't feel very strongly about this. 

Incidentally, even if we change this in the config file, still no need to change `RepoSecure` into two constructors of course.

23Skidoo replied:
> I wonder how many `local-repo`s are out there, if any at all.. 

Valid point.

> Could rename `remote-repo` to simply `repository`.. 

That'd be fine, I think. We can then deprecate `remote-repo` and `local-repo` config fields, and agree to remove them in some future release together with support for insecure repos. At that point, `RepoSecure` will be the only repo type in the `cabal-install` codebase, so it will be renamed to `Repo` or `Repository` as well.

edsko replied:
Note that HEAD already has a new kind of section; we now have, "old style" single line `remote-repo` definitions

```
remote-repo: hackage.haskell.org:http://hackage.haskell.org/packages/archive
```

as well as fancy new _sections_ (not yet available in any released version of `cabal-install`, AFAIK)

```
remote-repo hackage.haskell.org
  url: http://hackage.haskell.org/
  secure: True
  root-keys: 1035a452fd3ada87956f9e77595cfd5e41446781d7ba9ff9e58b94488ac0bad7
             89e692e45b53b575f79a02f82fe47359b0b577dec23b45d846d6e734ffcc887a
             dc4b6619e8ea2a0b72cad89a3803382f6acc8beda758be51660b5ce7c15e535b
  key-threshold: 2
```

we could, if we wanted, rename the latter to `repository`, and then slowly phase out both `local-repo` and `remote-repo`. Anyway, I don't feel too strongly about any of these, I'm just pointing out options.

23Skidoo replied:
Yep, that was basically what I had in mind (though perhaps I didn't communicate it very clearly) - rename new-style `remote-repo`  to `repository` and deprecate old-style `remote-repo` and `local-repo`.

23Skidoo commented on cabal-install/Distribution/Client/Setup.hs:
```diff
@@ -387,18 +399,96 @@ instance Monoid GlobalFlags where
+-- The security library has its own concept of a "local" repository, distinct
+-- from @cabal-install@'s; these are secure repositories, but live in the local
+-- file system. We use the convention that these repositories are identified by
+-- URLs of the form @file:/path/to/local/repo@.
```
Should probably be `file://` instead of `file:/`.

edsko replied:
The double slash in a URL comes before the host, and then a single slash after the host (the start of the path). Since there is no host, it should be a single slash. I didn't know this either until I tried to make this work with the `Network.URI.parseURI` function :)

23Skidoo replied:
Good to know. This is another quirk making life harder for users, a `path` field would be harder to get wrong.

dcoutts commented:
tar-0.4.4 released and #3028 ready.

edsko commented:
I've rebased this PR, but it really depends on 4 other, currently open, PRs (that's why there appear to be more commits here than there really are):

- [ ] #3028 (Unbundle tar)
- [ ] #3029 (Get rid of Repo construction from listBuildTreeRefs)
- [ ] #3030 (Introduce RepoContext)
- [ ] #3031 (Add support for request ehaders in getHttp)

I will rebase this again once those are all OKed and merged.

Then the following comments were made on this PR:

- [x] Make decision re bundling dependencies (decided: we will not bundle)
- [x] Release tar
- [x] Submit patch to http-client upstream (patch accepted and released)
- [ ] Release hackage-security
- [ ] Make decisions on which keys to bake in (if any)
- [x] Use HttpTransport in hackage-security
- [x] Some comments on HttpLib implementation (typo, export list); the HttpLib
      implementation now uses the HttpTransport; its export list is minimal.
- [x] Factor out `errMissingPackageList`
- [x] Factor out `errOutdatedPackageList`
- [x] Typos
- [x] Get rid of cabal's internal cache -- this is not done, but I've marked it
      here as "done" anyway because I'm no longer convinced that this is the
      right thing to do. Cabal will maintain it's own cache, with more info than
      the hackage-security cache; in fact, @hvr has even more extended cache.
      We do now however construct cabal's cache from the security cache, which
      should be significantly faster.
- [ ] Change the config style; should we rename the `remote-repo` section to
      `repository`, and ultimately phase out `remote-repo` and `local-repo`?
      Are we okay with specifying local repositories (in new layout) using
      `file:` URLs? This I haven't changed yet at all.

Note that the 'HttpTransport' and 'HttpLib' APIs don't line up _perfectly_; this is explained in more detail in the comments for `transportAdapter`; I've left https://github.com/well-typed/hackage-security/issues/42 open to remember this.

edsko commented:
Ok, I've released `hackage-security-0.5.0.1` which has some relaxed lower bounds, and pushed a fix to this branch, so that this now shows green on Travis for ghc 7.6, 7.8 and 7.10. It's still showing red on 7.4, but that's due to the `bytestring` issue.

dcoutts commented:
@23Skidoo @edsko re remote-repository. I agree, we should rename the new section name to just `repository`, and agree that at some point in the future we'll deprecate local-repo and remote-repo lines. Thus we'll also deprecate the old local repo format, and so local repos will use the same layout as the remote ones accesed via http. We'll also eventually want to get rid of the old code paths for non-secure local and remote repos.

edsko commented:
(Indeed, we could eventually remove all code paths except the ones that go through `hackage-security`; after all, we can set the key threshold to 0, so that people who want to set up non-secure repos would still need to add TUF metadata, but they can be trivial. Anyway, not really relevant now, just a thought.)

edsko commented:
@dcoutts , @23Skidoo so I've made a number of changes to the PR:

* The PR no longer bakes in any keys. We basically had only two options here: don't bake in any keys, or bake in the proper Hackage root keys (most of which we now have, though not yet all). However, the latter option means that we should then start using those root keys for Hackage as well. By not having any keys baked in we decouple these two things, and since this is opt-in for now anyway, we can just tell people how they should update their `~/.cabal/config` file. 
* I've renamed the `remote-repo` section to `repository`, as discussed. We can eventually deprecate `local-repo` and even `remote-repo`, but not yet.
* I've changed the `repoSecure` field to be a `Maybe`, and changed the config parsing/unparsing accordingly. This allows us to represent "use whatever the default is -- access securely or not" which is essential for a smooth transition from opt-in security to opt-out security. (Most people will have the `secure` value set to `Nothing`, whether they are using the old-style `remote-repo` field or the new-style `repository` section). If someone could take a quick look at my config-file stuff changes that would be good.

I think together with my earlier changes to this PR this addresses all concerns. Once #3028 and #3030 have been merged (and @dcoutts has released `tar` 0.5 that supports `bytestring-0.9`) this should be green on all versions of `ghc` and be ready for merging.

edsko commented:
Just for the record, the appropriate `repository` definition for the current Hackage is:

``` 
repository hackage.haskell.org
  url: http://hackage.haskell.org/
  secure: True
  root-keys: 89e692e45b53b575f79a02f82fe47359b0b577dec23b45d846d6e734ffcc887a
             dc4b6619e8ea2a0b72cad89a3803382f6acc8beda758be51660b5ce7c15e535b
             1035a452fd3ada87956f9e77595cfd5e41446781d7ba9ff9e58b94488ac0bad7
  key-threshold: 2
```

(This must now be specified manually even when using the `using-hackage-security` branch.)

23Skidoo commented:
OK, waiting on #3028 and #3030 to be merged.

@edsko 
> I've renamed the `remote-repo` section to `repository`, as discussed.
Have you also updated the docs?

edsko commented:
@23Skidoo I'm not sure exactly which documentation you are referring to. The man page, as far as I can see, does not give a definition of the format of the `~/.cabal/config` file, nor does that user manual. The man page _does_ mention the `remote-repo` key, but that still exists (as a command line flag, even); I'm not sure what the right fix there is.

edsko commented:
@23Skidoo Ok, I think this commit is good to go. The travis error on 7.10 I believe is completely unrelated to my patch; I've opened a separate ticket #3037 about this.

dcoutts commented:
LGTM.

23Skidoo commented:
@edsko 
> I'm not sure exactly which documentation you are referring to.

I mean the manual. If you don't know where to put it, you can add a section called "Secure repositories" or something. We can extend it to cover the rest of the config file later.

edsko commented:
@23Skidoo I'm happy enough to add something to the manual, but right now there isn't any section in the manual about configuration I think? Unless I'm totally blind :)

edsko commented:
@23Skidoo Does that need to happen before this can be merged?

23Skidoo commented:
@edsko Not really.

edsko commented:
@23Skidoo Then can I suggest we merge, and I promise to write the documentation section? Once we agree on where to put it :)

23Skidoo commented:
Sure, let's merge.

edsko commented:
Yay :tada: @23Skidoo I've opened a ticket on the hackage-security project for the docs https://github.com/well-typed/hackage-security/issues/148 . I will do this asap.
